fennel/datasets/datasets.py

```diff
@@ -2144,3 +2199,23 @@ def visitFirst(self, obj) -> DSSchema:
             )
         output_schema.name = f"'[Pipeline:{self.pipeline_name}]->first node'"
         return output_schema
+
+    def visitWindow(self, obj) -> DSSchema:
+        output_schema = copy.deepcopy(obj.dsschema())
+        output_schema_name = f"'[Pipeline:{self.pipeline_name}]->window node'"
+
+        if len(output_schema.keys) == 0:
+            raise ValueError(
+                f"`group_by` before `window` in {self.pipeline_name} must specify at least one key"
+            )
```

Comment:
Are we going to force that window must be preeceded by a groupby? I think that's unnecessary as long as it has some key columns. In that case, the error message should not refer to groupby. 

Also IIRC, groupby itself will fail if it had no keys.

Reply:
I don't think groupby itself fails currently. Also the window operator I thought is similar to first, so made the code in such a way that window comes after groupby. If we want other way also I can do that, but if we do then I think we should also make first also similar. @aditya-nambiar your thoughts on this?

Reply:
@nikhilgarg28 , currently we require aggregate and first to be preceded by groupby and we could follow similar suite for now, later we can refactor the code to make them independent, but lets proceed to keep the code change contained.